core/src/main/java/com/linecorp/armeria/client/IgnoreHostsTrustManager.java

```diff
@@ -0,0 +1,115 @@
+ *  License for the specific language governing permissions and limitations
+ *  under the License.
+ */
+
+package com.linecorp.armeria.client;
+
+import static java.util.Objects.requireNonNull;
+
+import java.net.Socket;
+import java.security.GeneralSecurityException;
+import java.security.KeyStore;
+import java.security.cert.CertificateException;
+import java.security.cert.X509Certificate;
+import java.util.Set;
+
+import javax.net.ssl.SSLEngine;
+import javax.net.ssl.TrustManager;
+import javax.net.ssl.TrustManagerFactory;
+import javax.net.ssl.X509ExtendedTrustManager;
+
+import com.google.common.collect.ImmutableSet;
+
+/**
+ * An implementation of {@link X509ExtendedTrustManager} that skips verification on whitelisted hosts.
+ */
+final class IgnoreHostsTrustManager extends X509ExtendedTrustManager {
+
+    private final X509ExtendedTrustManager delegate;
+    private final Set<String> insecureHosts;
+
+    IgnoreHostsTrustManager(X509ExtendedTrustManager delegate, Set<String> insecureHosts) {
+        this.delegate = delegate;
+        this.insecureHosts = insecureHosts;
+    }
+
+    /**
+     * Returns new {@link IgnoreHostsTrustManager} instance.
+     */
+    static IgnoreHostsTrustManager of(String... insecureHosts) {
+        X509ExtendedTrustManager delegate = null;
+        try {
+            final TrustManagerFactory trustManagerFactory = TrustManagerFactory
+                    .getInstance(TrustManagerFactory.getDefaultAlgorithm());
+            trustManagerFactory.init((KeyStore) null);
+            final TrustManager[] trustManagers = trustManagerFactory.getTrustManagers();
+            for (TrustManager tm : trustManagers) {
+                if (tm instanceof X509ExtendedTrustManager) {
+                    delegate = (X509ExtendedTrustManager) tm;
+                    break;
+                }
+            }
+        } catch (GeneralSecurityException ignored) {
+            // ignore
+        }
+        requireNonNull(delegate, "cannot resolve default trust manager");
+        return new IgnoreHostsTrustManager(delegate, ImmutableSet.copyOf(insecureHosts));
+    }
+
+    @Override
+    public void checkServerTrusted(X509Certificate[] x509Certificates, String s, Socket socket)
+            throws CertificateException {
+        if (!insecureHosts.contains(socket.getInetAddress().getHostName())) {
+            delegate.checkServerTrusted(x509Certificates, s, socket);
+        }
+    }
+
+    @Override
+    public void checkServerTrusted(X509Certificate[] x509Certificates, String s, SSLEngine sslEngine)
+            throws CertificateException {
+        if (!insecureHosts.contains(sslEngine.getPeerHost())) {
+            delegate.checkServerTrusted(x509Certificates, s, sslEngine);
+        }
+    }
+
+    @Override
+    public void checkServerTrusted(X509Certificate[] x509Certificates, String s)
+            throws CertificateException {
+        throw new UnsupportedOperationException();
```

Comment:
It does not give access to the hostname so I wasn't sure if it should be implemented.

Reply:
It's probably fine not to implement this method and the method that accepts `Socket` because we always use an `SSLEngine` for TLS communication. If we have to implement this, we could retrieve the host name from `x509Certificates[0]`.

Reply:
Thanks for chiming in. I'm not entirely familiar with TLS so advice is much appreciated here. But it looks to me that we don't need this right now, so maybe it can be added down the road if needed?

Reply:
LGTM. 👍